TFNetworkRecLayer.py

```diff
@@ -4428,6 +4430,22 @@ def transform_config_dict(cls, d, network, get_layer):
         [get_layer(name) for name in d["explicit_search_sources"]] if network.search_flag else [])
     super(ChoiceLayer, cls).transform_config_dict(d, network=network, get_layer=get_layer)
 
+  @classmethod
+  def _create_search_beam(cls, name, beam_size, sources, network):
+    """
+    :param str name:
+    :param int beam_size:
+    :param list[LayerBase] sources:
+    :param TFNetwork.TFNetwork network:
+    """
+    from TFUtil import SearchBeam
+    search_dep = NotSpecified
+    if sources and sources[0] and not sources[0].output.undefined:
+      search_dep = sources[0].output.beam
+    return SearchBeam(
+      beam_size=beam_size, dependency=search_dep,
+      name="%s%s" % (network.get_absolute_name_prefix(), name))
```

Comment:
This has still the hardcoded solution. This is not nice and can potentially break later.

Reply:
Uhm, what exactly do you mean is hardcoded here? The fact that the same SearchBeam is created twice, relying on the name? 

Well. the idea of `get_sub_layer_out_data_from_opts` was to create `out_data` of the sub-layer without depending on the `out_data` or an instance of the parent layer. Do you want to change that?

Another option would be to create the parent layer beam by calling `get_out_data_from_opts` using the `parent_layer_kwargs` and then just use `out_data.beam` and discard the rest. But this is the most generic you could do here, right?

Reply:
No, for `name` you assume that it matches to other code.

Reply:
I'm still not sure I get what you mean. An instance of SearchBeam is characterised by the owner layer name and comparison of beams relies on the name, this is nothing I introduced. Also I did not change the name format `"%s%s" % (network.get_absolute_name_prefix(), name))`.
All I need to do is to use `parent_layer_kwargs["name"]` instead of `layer_name` when creating the beam of the sub layer. How does this depend on other code? Well, I'm assuming that the parent layer beam has indeed the name of the parent layer, but that's it.

Reply:
You assume that we have this specific name format.

Reply:
Yes, because it was there before ':D As of now, it is a local property of ChoiceLayer. And the class methods have access to it.
I want to go from kwargs to SearchBeam. How would I do this without using some kind of naming convention?

Reply:
I'm just saying that you have this assumption here about the name format, and this could break at some point. Better write code in a way that it will not break.

Reply:
I really don't understand what my PR changed about the naming. As before, there is one, and only one, place in the code that *defines* the naming of SearchBeams.
Again, how do I get from kwargs to SearchBeam? In kwargs there is no SearchBeam, so I have to create one. For this I need a name.

Reply:
I thought there are two places now? Where was the `SearchBeam` created before?